controller/workflow.go

```diff
@@ -2,39 +2,67 @@ package controller
 
 import (
 	"context"
+	"errors"
 	"sync"
 
+	"golang.org/x/sync/errgroup"
+
 	"github.com/kuadrant/policy-machinery/machinery"
 )
 
 // Workflow runs an optional precondition reconciliation function, then dispatches the reconciliation event to
 // a list of concurrent reconciliation tasks, and runs an optional postcondition reconciliation function.
+//
+// If any of the reconciliation functions returns an error, the error is handled by an optional error handler.
+// The error passed to the error handler func is conflated with any ocasional error carried over into the call
+// to the workflow in the first place. It is up to the error handler to decide how to handle the error and whether
+// to supress it or raise it again. Supressed errors cause the workflow to continue running, while raised errors
+// interrupt the workflow. If the error handler is nil, the error is raised.
 type Workflow struct {
 	Precondition  ReconcileFunc
 	Tasks         []ReconcileFunc
 	Postcondition ReconcileFunc
+	ErrorHandler  ReconcileFunc
```

Comment:
I am not sure I like this. Currently it is a gut feeling but seems complicated from the user prospective. Not sure what would be better, or even worse.

Reply:
I referred to the use case for this as Problem 2 in [this comment](https://github.com/Kuadrant/policy-machinery/pull/31#issuecomment-2358443009).

In fact, users "start off loading the errors to upper layers" is indeed the capability I wanted to enable with this feature. I guess I don't see this inherently as a bad thing. There are cases where you want to do it and cases where you don't.

Do you think there is no case where this is a good thing, @Boomatang? I'm honestly asking it.

Reply:
Of course there will be case where this is a good thing. This is not at about this being good or bad, it's about it being right. Every example that I can think of that would use this I can see being solved without using it. Yes, some would require other work to be complete, but that is here nor there. The short circuit of a workflow if there is an error in the Precondition is the only part that I can't seem to see an alternative too. If there is an error in one routine in the Tasks list I would still expect the Post-condition to run.  

I am happy to accept this and see how it works in practices. The library is no where near V1 and there is not a lot of users yet. I would like to see some other alternatives explored and seen this being used with in Kuadrant operator will help narrow down what is needed.